src/awkward/_v2/cpp-headers/GrowableBuffer.h

```diff
@@ -0,0 +1,263 @@
+    /// This is similar to NumPy's
+    /// [arange](https://docs.scipy.org/doc/numpy/reference/generated/numpy.arange.html).
+    static GrowableBuffer<PRIMITIVE>
+    arange(size_t initial, size_t length) {
+      size_t actual = initial;
+      if (actual < length) {
+        actual = length;
+      }
+      auto ptr = std::unique_ptr<PRIMITIVE>(new PRIMITIVE[actual]);
+      PRIMITIVE* rawptr = ptr.get();
+      for (size_t i = 0;  i < length;  i++) {
+        rawptr[i] = (PRIMITIVE)i;
+      }
+      return GrowableBuffer(initial, std::move(ptr), length, actual);
+    }
+
+    /// @brief Creates a GrowableBuffer from a full set of parameters.
+    ///
+    /// @param initial Initial size configuration for building a panel.
+    /// @param ptr Reference-counted pointer to the array buffer.
+    /// @param length Currently used number of elements.
+    /// @param reserved Currently allocated number of elements.
+    ///
+    /// Although the #length increments every time #append is called,
+    /// it is always less than or equal to #reserved because of
+    /// allocations of new panels.
+    GrowableBuffer(size_t initial,
+                   std::unique_ptr<PRIMITIVE> ptr,
+                   size_t length,
+                   size_t reserved)
+        : initial_(initial) {
+      ptr_.push_back(std::move(ptr));
+      length_.push_back(length);
+      reserved_.push_back(reserved);
+    }
+
+    /// @brief Creates a GrowableBuffer by allocating a new buffer, taking an
+    /// initial #reserved from #initial.
+    GrowableBuffer(size_t initial)
+        : GrowableBuffer(initial,
+                         std::unique_ptr<PRIMITIVE>(new PRIMITIVE[initial]),
+                         0,
+                         initial) { }
+
+    /// @brief Currently used number of elements.
+    ///
+    /// Although the #length increments every time #append is called,
+    /// it is always less than or equal to #reserved because of
+    /// allocations of new panels.
+    size_t
+    length() const {
+      return std::accumulate(length_.begin(), length_.end(), (size_t)0);
+    }
+
+    /// @brief Currently allocated number of elements.
+    ///
+    /// Although the #length increments every time #append is called,
+    /// it is always less than or equal to #reserved because of
+    /// allocations of new panels.
+    size_t
+    reserved() const {
+      return std::accumulate(reserved_.begin(), reserved_.end(), (size_t)0);
+    }
+
+    /// @brief Discards accumulated data, the #reserved returns to
+    /// initial, and a new #ptr is allocated.
+    void
+    clear() {
+      length_.clear();
+      length_.push_back(0);
+      reserved_.clear();
+      reserved_.push_back(initial_);
+      ptr_.clear();
+      ptr_.push_back(std::unique_ptr<PRIMITIVE>(new PRIMITIVE[initial_]));
+    }
+
+    /// @brief Inserts one `datum` into the panel.
+    void
+    fill_panel(PRIMITIVE datum) {
+      if (length_[ptr_.size()-1] < reserved_[ptr_.size()-1]) {
```

Comment:
Should this member function be private? Also, this if statement looks like a double check. Should it be removed?